lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +802,460 @@ def __str__(self):
+        if ref_type not in self.known_types():
+            # This is made a TypeError and not an InvalidGitRef because it cannot arise
+            # from invalid user input.
+            raise TypeError('GitRef can only have the types {0}, but was given {1!r}'
+                            .format(self.known_types(), ref_type))
+        self._ref_type = ref_type
+        if not isinstance(ref, six.string_types):
+            # This case can arise from user error when editing package.py files, so we
+            # want to catch this and handle it specially to provide more context.
+            raise InvalidGitRef('git reference was not a string: {0!r}'.format(ref))
+        self._ref = ref
+
+    def __repr__(self):
+        return '{0}(ref_type={1!r}, ref={2!r})'.format(type(self).__name__,
+                                                       self.ref_type, self._ref)
+
+    def __str__(self):
+        return '<{0}>'.format(self.repo_info_for_reference())
+
+    # Matches any 7-40 character hexadecimal string.
+    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
+
+    @classmethod
+    def commit(cls, ref):
+        # type: (str) -> GitRef
+        if not isinstance(ref, six.string_types):
+            # The exception message from `re.match()` does not print the value when
+            # a non-string argument is provided, so we special-case that check here.
+            raise InvalidGitRef('git reference was not a string: {0}'.format(ref))
+        if not cls._commit_rx.match(ref):
+            raise InvalidGitRef(dedent("""\
+            Spack requires that references to individual git commits be specified via
+            a 7-40 character hexadecimal string, but received {0!r} instead.
+
+            A valid commit hash case-insensitively matches the regular expression '{1}'.
+            7 hex characters is the size printed out by `git log --format='%h'`, while
+            40 hex characters is the size printed out by `git log --format='%H'`.
+            """.format(ref, cls._commit_rx.pattern)))
+        return cls('commit', ref)
+
+    @classmethod
+    def tag(cls, ref):
+        # type: (str) -> GitRef
+        return cls('tag', ref)
+
+    @classmethod
+    def branch(cls, ref):
+        # type: (str) -> GitRef
+        return cls('branch', ref)
+
+    @classmethod
+    def from_version_directive(cls, kwargs):
+        # type: (Any) -> GitRef
+        # (1) Extract the relevant arguments.
+        commit = kwargs.pop('commit', None)
+        tag = kwargs.pop('tag', None)
+        branch = kwargs.pop('branch', None)
+        version_name = kwargs.pop('version_name', None)
+        # (2) Ensure no mutually exclusive kwargs are provided.
+        num_specified = len(list(filter(None, [commit, tag, branch])))
+        # (3) If no explicit git ref arguments are provided, assume that the version
+        # string itself is to be used as the branch name.
+        if (num_specified == 0) and version_name is not None:
```

Comment:
I think if this went through a Python brain linter some folks would say you can just do `if not num_specified` instead of explicitly checking for 0. But this works okay too. I don't think you need the extra parentheses.

Reply:
I have been burned by python conditional syntax one too many times to feel comfortable about removing the parentheses, since `and` has surprisingly high precedence (or maybe it's `or`?)! The `== 0` is also a workaround for that, since I don't want someone reviewing this to have to think too hard to make sure it's validating things correctly, especially if this code gets edited later!

I am willing to change this (I believe you are correct that `not` has higher precedence than `and`), but I would really prefer to keep this the way it is as a form of documentation for the intent of the code.

Reply:
Sure! I think I was just wondering why not do:

```python
if not num_specified and not version_name
```
We know num_specified is going to be a positive number or 0, and is a empty string version name allowed?

No worries if you don't want to change - what you have now is actually more readable. But I wanted to check that version_name isn't an empty string. Sometimes I often do:

```
version_name = version_name or None
version_name = version_name or ""
```
to make sure it's the exact empty-thing that I want :)

Reply:
This was a toughie! Thank you for writing out all your reasoning!

In this case I explicitly wanted to avoid `version_name` being a blank string (which I'd *totally* failed to consider until you mentioned it!!), so I converted this to:
```python
        if (num_specified == 0) and version_name:
            return cls.Branch(version_name)
````

I've kept `(num_specified == 0)` because I'm e.g. comparing against `num_specified == 2` later and it feels more appropriate to maintain the same structure. I agree that it's overparenthesized but I think in this case it could be genuinely confusing to see the alternative:
```python
        if not num_specified and version_name:
            return cls.Branch(version_name)